File: julia/superReducedString.jl
# link = https://www.hackerrank.com/challenges/reduced-string/problem

function superReducedString(inp)
    empty = "Empty String"
    s = split(inp, "")
    n = length(s)
    
    x = 2
    while x <= n
        if x <= 1 x += 1 end
        if s[x-1] == s[x]
            deleteat!(s, (x-1, x))
            n -= 2
            x -= 2
        end
        x += 1
    end
    n <= 0 ? empty : join(s)
end

function main()
    inp = readline()
    superReducedString(inp) |> println
end
main()

File: cpp/superReducedString.cpp
// link = https://www.hackerrank.com/challenges/reduced-string/problem
#include <bits/stdc++.h>

using namespace std;

// Complete the superReducedString function below.
string superReducedString(string s) {
    string empty = "Empty String";
    int n = s.size();
    for(int x = 1; x < n; ++x){
        if(s[x-1] == s[x]){ // remove both
            s.erase(s.begin() + x - 1, s.begin() + x + 1);
            // decrease n
            n -= 2;
            x -= 2;
        }
    }
    return (n > 0 ? s : empty);
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string s;
    getline(cin, s);

    string result = superReducedString(s);

    fout << result << "\n";

    fout.close();

    return 0;
}


/* REGEX SOLUTION
string superReducedString(string s) {
    std::regex dupe("(.)\\1");
    
    for(int x{}; x < 100; ++x)
        s = std::regex_replace(s, dupe, "");
    return (s.size() <= 0 ? "Empty String" : s);
}
*/
